migrations/versions/20181212_rename_page_table.py

```diff
@@ -0,0 +1,160 @@
+    )
+
+    op.alter_column("page", "id", nullable=False)
+    op.alter_column("data_source_in_page", "measure_version_id", nullable=False)
+    op.alter_column("upload", "measure_version_id", nullable=False)
+    op.alter_column("dimension", "measure_version_id", nullable=False)
+
+    # * rename tables:
+    op.rename_table("page", "measure_version")
+    op.rename_table("data_source_in_page", "data_source_in_measure_version")
+
+    # * add constraints back
+    # * Add foreign key and not nullable constraints to `data_source_in_page`, `upload`, `dimension`
+    op.create_primary_key(op.f("measure_version_id_pkey"), "measure_version", ["id", "guid", "version"])
+    op.create_primary_key(
+        op.f("data_source_in_measure_version_data_source_id_pkey"),
+        "data_source_in_measure_version",
+        ["data_source_id", "measure_version_id"],
+    )
+
+    op.create_foreign_key(
+        op.f("measure_version_parent_id_fkey"),
+        "measure_version",
+        "measure_version",
+        ["parent_id", "parent_guid", "parent_version"],
+        ["id", "guid", "version"],
+    )
+    op.create_foreign_key(
+        op.f("upload_measure_version_id_fkey"),
+        "upload",
+        "measure_version",
+        ["measure_version_id", "page_id", "page_version"],
+        ["id", "guid", "version"],
+    )
+    op.create_foreign_key(
+        op.f("dimension_measure_version_id_fkey"),
+        "dimension",
+        "measure_version",
+        ["measure_version_id", "page_id", "page_version"],
+        ["id", "guid", "version"],
+    )
+
+    op.create_foreign_key(
+        op.f("data_source_in_measure_version_measure_version_id_fkey"),
+        "data_source_in_measure_version",
+        "measure_version",
+        ["measure_version_id", "page_guid", "page_version"],
+        ["id", "guid", "version"],
+    )
+
+
+def downgrade():
+    op.drop_constraint(
+        "data_source_in_measure_version_measure_version_id_fkey", "data_source_in_measure_version", type_="foreignkey"
+    )
+    op.drop_constraint("dimension_measure_version_id_fkey", "dimension", type_="foreignkey")
+    op.drop_constraint("upload_measure_version_id_fkey", "upload", type_="foreignkey")
+    op.drop_constraint("measure_version_parent_id_fkey", "measure_version", type_="foreignkey")
+    op.drop_constraint(
+        "data_source_in_measure_version_data_source_id_pkey", "data_source_in_measure_version", type_="primary"
+    )
+    op.drop_constraint("measure_version_id_pkey", "measure_version", type_="primary")
+
+    op.rename_table("measure_version", "page")
+    op.rename_table("data_source_in_measure_version", "data_source_in_page")
+
+    op.drop_column("dimension", "measure_version_id")
+    op.drop_column("upload", "measure_version_id")
+    op.drop_column("data_source_in_page", "measure_version_id")
+    op.drop_column("page", "id")
```

Comment:
Not sure it will let you drop `id` before `parent_id`? Or is it OK because the constraint has gone already?

Still, it's nice to have the `downgrade()` in exact reverse order of `upgrade()`, which this isn't quite.

Reply:
Yep, once the FK constraint is gone, it lets you drop, so this downgrade works. But I'll go and put them in the same order regardless - as you say, it's nice for it to be consistently reversed.